services/src/main/java/org/keycloak/broker/saml/SAMLIdentityProvider.java

```diff
@@ -369,9 +384,56 @@ public Response export(UriInfo uriInfo, RealmModel realm, String format) {
                         }
                     });
 
-            String descriptor = SPMetadataDescriptor.getSPDescriptor(authnBinding, endpoint, endpoint,
-              wantAuthnRequestsSigned, wantAssertionsSigned, wantAssertionsEncrypted,
-              entityId, nameIDPolicyFormat, signingKeys, encryptionKeys);
+            // Prepare the metadata descriptor model
+            StringWriter sw = new StringWriter();
+            XMLStreamWriter writer = StaxUtil.getXMLStreamWriter(sw);
+            SAMLMetadataWriter metadataWriter = new SAMLMetadataWriter(writer);
+
+            EntityDescriptorType entityDescriptor = SPMetadataDescriptor.buildSPdescriptor(
+                authnBinding, authnBinding, endpoint, endpoint,
+                wantAuthnRequestsSigned, wantAssertionsSigned, wantAssertionsEncrypted,
+                entityId, nameIDPolicyFormat, signingKeys, encryptionKeys);
+
+            // Create the AttributeConsumingService
+            AttributeConsumingServiceType attributeConsumingService = new AttributeConsumingServiceType(attributeConsumingServiceIndex);
+            attributeConsumingService.setIsDefault(true);
+
+            if (attributeConsumingServiceName != null && attributeConsumingServiceName.length() > 0)
+            {
+                String currentLocale = realm.getDefaultLocale() == null ? "en": realm.getDefaultLocale();
+                LocalizedNameType attributeConsumingServiceNameElement = new LocalizedNameType(currentLocale);
+                attributeConsumingServiceNameElement.setValue(attributeConsumingServiceName);
+                attributeConsumingService.addServiceName(attributeConsumingServiceNameElement);
+            }
+
+            // Look for the SP descriptor and add the attribute consuming service
+            for (EntityDescriptorType.EDTChoiceType choiceType: entityDescriptor.getChoiceType()) {
+                List<EntityDescriptorType.EDTDescriptorChoiceType> descriptors = choiceType.getDescriptors();
+
+                if (descriptors != null) {
+                    for (EntityDescriptorType.EDTDescriptorChoiceType descriptor: descriptors) {
+                        if (descriptor.getSpDescriptor() != null) {
+                            descriptor.getSpDescriptor().addAttributeConsumerService(attributeConsumingService);
+                        }
+                    }
+                }
+            }
+            
+            // Add the attribute mappers
+            realm.getIdentityProviderMappersByAliasStream(getConfig().getAlias())
+                .forEach(mapper -> {
+                    IdentityProviderMapper target = (IdentityProviderMapper) session.getKeycloakSessionFactory().getProviderFactory(IdentityProviderMapper.class, mapper.getIdentityProviderMapper());
```

Comment:
Can we break this lambda into more calls of `map` and `filter`?

Reply:
I'm not really fluent with Java streams, but I tried and couldn't make it work because the `updateMetadata` call requires both the interface instance and the mapper model:

```
realm.getIdentityProviderMappersByAliasStream(getConfig().getAlias())
  .map(mapper -> (IdentityProviderMapper) session.getKeycloakSessionFactory().getProviderFactory(IdentityProviderMapper.class, mapper.getIdentityProviderMapper()))
  .filter(target -> target instanceof SamlMetadataDescriptorUpdater)
  .map(target -> (SamlMetadataDescriptorUpdater)target)
  .forEach(metadataAttrProvider -> metadataAttrProvider.updateMetadata(mapper, entityDescriptor)); << error here, mapper is unknown  
```

I could probably change the last map call to return a pair-like object, but it seems much harder to read than the original foreach. WDYT?

Reply:
I see. I think you are right, this way it is easier to read. I just don't like that the lambda is quite long, but I don't see a better option.